### PR TITLE
Add `extract`

### PR DESCRIPTION
**Pull request by Harshan01**
This PR is for the feature request Issue #2223

### CODE REVIEW

**Comment by Harshan01** [comment]
Though numpy mentions that the `condition` should be of the same size as the input array, it works for all sizes which are less than that of the array. So, I am not sure if tests have to be added for empty array cases and cases where `condition` is of smaller size than input array.

**Comment by Harshan01** [comment]
Presently, I made sure that the input array has to be `cupy.ndarray`. I would like to receive suggestions if the function should also accept python list and numpy array as input array.

File: tests/cupy_tests/indexing_tests/test_indexing.py [hunk 1]
```diff
@@ -93,6 +93,21 @@ def test_diagonal_invalid2(self, xp):
...
+                      [True, False, True]])
+        return xp.extract(b, a)
+
+    @testing.for_signed_dtypes()
+    @testing.numpy_cupy_array_equal()
+    def test_extract_no_bool(self, xp, dtype):
+        a = testing.shaped_arange((3, 3), xp)
+        b = xp.array([[1, 0, 1], [0, 1, 0], [1, 0, 1]], dtype=dtype)
+        return xp.extract(b, a)
+
```
Comment (asi1024) [inline tests/cupy_tests/indexing_tests/test_indexing.py]: Please add tests for `a.shape != b.shape` and `a.size != b.size`

File: cupy/indexing/indexing.py [hunk 2]
```diff
@@ -104,4 +104,38 @@ def diagonal(a, offset=0, axis1=0, axis2=1):
...
+            This function may synchronize the device.
+
+    .. seealso:: :func:`numpy.extract`
+    """
+
+    if not isinstance(a, cupy.ndarray):
+        raise TypeError('extract requires input array to be cupy.ndarray')
+
+    if not isinstance(condition, cupy.ndarray):
+        condition = cupy.array(condition)
```
Comment (asi1024) [inline cupy/indexing/indexing.py]: The above docstring says that the argument `condition` is required to be of `cupy.ndarray`.

Comment (jakirkham) [reply cupy/indexing/indexing.py]: Well NumPy also allows this.

```python
In [1]: import numpy                                                            

In [2]: a = numpy.arange(6).reshape(2, 3)                                       

In [3]: a                                                                       
Out[3]: 
array([[0, 1, 2],
       [3, 4, 5]])

In [4]: numpy.extract([[True, False, True], 
   ...:                [False, True, True]], 
   ...:               a)                                                        
Out[4]: array([0, 2, 4, 5])
```

Though perhaps the question is how do we specify something that is `ndarray`-like in a way that is compatible with CuPy's documentation practices 🙂

Comment (Harshan01) [reply cupy/indexing/indexing.py]: I too had the same doubt. There was no mention of `array_like` in CuPy's documentations.

**Comment by Harshan01** [comment]
@asi1024 I have made the requested changes. Can you please review them?

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM. Thanks for your contribution!

**Pull request merged**